ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java

```diff
@@ -473,9 +477,20 @@ private boolean containsHostFromHostGroups(String configType, String propertyNam
 
             // set the properties what configure the NameNode Active/Standby status for each nameservice
             if (!activeNameNodeHostnames.isEmpty() && !standbyNameNodeHostnames.isEmpty()) {
-              clusterConfig.setProperty("hadoop-env", HDFS_ACTIVE_NAMENODE_SET_PROPERTY_NAME, String.join(",", activeNameNodeHostnames));
-              clusterConfig.setProperty("hadoop-env", HDFS_STANDBY_NAMENODE_SET_PROPERTY_NAME, String.join(",", standbyNameNodeHostnames));
-              configTypesUpdated.add("hadoop-env");
+              clusterConfig.setProperty(HADOOP_ENV_CONFIG_TYPE_NAME, HDFS_ACTIVE_NAMENODE_SET_PROPERTY_NAME, String.join(",", activeNameNodeHostnames));
+              clusterConfig.setProperty(HADOOP_ENV_CONFIG_TYPE_NAME, HDFS_STANDBY_NAMENODE_SET_PROPERTY_NAME, String.join(",", standbyNameNodeHostnames));
+
+              // also set the clusterID property, required for Federation installs of HDFS
+              if (!isPropertySet(clusterProps, HADOOP_ENV_CONFIG_TYPE_NAME, HDFS_HA_INITIAL_CLUSTER_ID_PROPERTY_NAME)) {
+                final String clusterName = getClusterName();
+                if (clusterName != null) {
+                  clusterConfig.setProperty(HADOOP_ENV_CONFIG_TYPE_NAME, HDFS_HA_INITIAL_CLUSTER_ID_PROPERTY_NAME, clusterName);
+                } else {
+                  LOG.warn("Cluster name could not obtained, this may indicate a deployment or configuration error.");
+                }
```

Comment:
Shouldn't missing cluster name mean an exception or are there valid cases when cluster name could be missing?

Reply:
That's a good question.  I've checked the usages of the getClusterName() method throughout the server code, and some cases treat this as an exception case, and some just proceed assuming that the name is always non-null.

I don't think there are any valid cases when the cluster name would be missing, so I'll update this code to throw an exception in the event that a null cluster name is found.

Reply:
Updated the patch to include more error-checking in this area, and added a unit test to verify that the expected exception is thrown in the event that getClusterName() returns null.